pkg/allocateip/allocateip.go

```diff
@@ -326,21 +487,38 @@ func removeHostTunnelAddr(ctx context.Context, c client.Interface, nodename stri
 		}
 
 		// Determine if we need to do any work.
-		ipipTunnelAddrExists := (node.Spec.BGP != nil && node.Spec.BGP.IPv4IPIPTunnelAddr != "")
+		ipipTunnelAddrExists := node.Spec.BGP != nil && node.Spec.BGP.IPv4IPIPTunnelAddr != ""
 		vxlanTunnelAddrExists := node.Spec.IPv4VXLANTunnelAddr != ""
-		if (vxlan && !vxlanTunnelAddrExists) || (!vxlan && !ipipTunnelAddrExists) {
+		wireguardTunnelAddrExists := node.Spec.Wireguard != nil && node.Spec.Wireguard.InterfaceIPv4Address != ""
+		if (attrType == ipam.AttributeTypeVXLAN && !vxlanTunnelAddrExists) ||
+			(attrType == ipam.AttributeTypeIPIP && !ipipTunnelAddrExists) ||
+			(attrType == ipam.AttributeTypeWireguard && !wireguardTunnelAddrExists) {
 			logCtx.Debug("No tunnel address assigned, and not required")
 			return
 		}
 
 		// Find out the currently assigned address and remove it from the node.
 		var ipAddr *net.IP
-		if vxlan {
+		switch attrType {
+		case ipam.AttributeTypeVXLAN:
 			ipAddr = net.ParseIP(node.Spec.IPv4VXLANTunnelAddr)
 			node.Spec.IPv4VXLANTunnelAddr = ""
-		} else if node.Spec.BGP != nil {
-			ipAddr = net.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
-			node.Spec.BGP.IPv4IPIPTunnelAddr = ""
+		case ipam.AttributeTypeIPIP:
+			if node.Spec.BGP != nil {
+				ipAddr = net.ParseIP(node.Spec.BGP.IPv4IPIPTunnelAddr)
+				node.Spec.BGP.IPv4IPIPTunnelAddr = ""
+			}
+		case ipam.AttributeTypeWireguard:
+			if node.Spec.Wireguard != nil {
+				ipAddr = net.ParseIP(node.Spec.Wireguard.InterfaceIPv4Address)
+				node.Spec.Wireguard.InterfaceIPv4Address = ""
+			}
+		}
+
+		// If no IP address found then there is nothing to remove.
```

Comment:
I don't think this is correct - it's possible that the node and IPAM data is out of sync. We should attempt to remove the address every time no matter what rather than short circuit here, otherwise we might leak IP addresses. 

We release by handle so that we don't need the address in hand in order to do the release.

Reply:
Hrm, maybe that's not true in this code path. Seems like we assume the IP is non-nil below...

Reply:
Hmmm, not sure why I added that - certainly didn't mean to change the existing behavior.  My bad!

Reply:
I think that is a bug, though. Even if there is no IP on the node, we should try to release it by handle anyway to clean up anything that might be left over in IPAM.